pkg/tracing/tracer.go

```diff
@@ -0,0 +1,277 @@
+
+	jobs       map[EventType]chan *Job
+	jobsClosed sync2.AtomicBool
+	rpcWg      sync.WaitGroup
+}
+
+// NewTracer creates a new Tracer
+func NewTracer(cfg Config) *Tracer {
+	t := Tracer{
+		cfg:   cfg,
+		tso:   newTsoGenerator(),
+		idGen: NewIDGen(),
+	}
+	t.jobsClosed.Set(true)
+	t.ctx, t.cancel = context.WithCancel(context.Background())
+	return &t
+}
+
+// Enable returns whether tracing is enabled
+func (t *Tracer) Enable() bool {
+	return t.cfg.Enable
+}
+
+// Start starts tracing service
+func (t *Tracer) Start() {
+	if !t.Enable() {
+		return
+	}
+	conn, err := grpc.Dial(t.cfg.TracerAddr, grpc.WithInsecure(), grpc.WithBackoffMaxDelay(3*time.Second))
+	if err != nil {
+		log.Errorf("[tracer] grpc dial error: %s", errors.ErrorStack(err))
+		return
+	}
+	t.cli = pb.NewTracerClient(conn)
+	t.closed.Set(false)
+	t.newJobChans()
+
+	t.wg.Add(1)
+	go func() {
+		ctx2, _ := context.WithCancel(t.ctx)
+		t.tsoProcessor(ctx2)
+		// cancel()
+	}()
+
+	for _, ch := range t.jobs {
+		t.wg.Add(1)
+		go func(c <-chan *Job) {
+			ctx2, _ := context.WithCancel(t.ctx)
+			t.jobProcessor(ctx2, c)
+			// cancel()
+		}(ch)
+	}
+}
+
+// Stop stops tracer
+func (t *Tracer) Stop() {
+	if !t.Enable() {
+		return
+	}
+	t.Lock()
+	defer t.Unlock()
+	if t.closed.Get() {
+		return
+	}
+	if t.cancel != nil {
+		t.cancel()
+	}
+	t.closed.Set(true)
+	t.wg.Wait()
+}
+
+// PrepareRPC calls before each rpc request to tracing server
+func (t *Tracer) PrepareRPC() {
+	t.rpcWg.Add(1)
+}
+
+func (t *Tracer) tsoProcessor(ctx context.Context) {
+	defer t.wg.Done()
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-time.After(1 * time.Minute):
+			t.PrepareRPC()
+			err := t.syncTS()
+			if err != nil {
+				log.Errorf("[tracer] sync timestamp error: %s", errors.ErrorStack(err))
+			}
+		}
+	}
+}
+
+// Assume we send GetTSO with local timestamp t1, timestamp gap between local
+// and remote tso service is tx, remote tso service get GetTSO at timestamp t2
+// and send response back as soon as possible. We get RPC response at local
+// timestamp t3. then we have t1 + tx + ttl = t2; t2 - tx + ttl = t3
+// so the real sending RPC request timestamp is t1 + tx = t2 + t1/2 - t3/2
+func (t *Tracer) syncTS() error {
+	defer t.rpcWg.Done()
+	t.tso.Lock()
+	defer t.tso.Unlock()
+	t.tso.localTS = time.Now().UnixNano()
+	req := &pb.GetTSORequest{Id: t.cfg.Source}
+	resp, err := t.cli.GetTSO(t.ctx, req)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if !resp.Result {
+		return errors.Errorf("sync ts with error: %s", resp.Msg)
+	}
+	currentTS := time.Now().UnixNano()
+	oldSyncedTS := t.tso.syncedTS
+	t.tso.syncedTS = resp.Ts + t.tso.localTS/2 - currentTS/2
+	log.Debugf("[tracer] syncedTS from %d to %d, localTS %d", oldSyncedTS, t.tso.syncedTS, t.tso.localTS)
+	return nil
+}
+
+func (t *Tracer) newJobChans() {
+	t.closeJobChans()
+	t.jobs = make(map[EventType]chan *Job)
+	for i := EventSyncerBinlog; i < EventFlush; i++ {
+		t.jobs[i] = make(chan *Job, t.cfg.BatchSize+1)
```

Comment:
why `+1`?

Reply:
not needed, change buffer length to `t.cfg.BatchSize`